Enable API chunking and promote to beta for 1.9

smarterclayton opened the pull request:
All list watchers default to using chunking.  The server by default fills pages to avoid low cardinality filters from making excessive numbers of requests.  Fix an issue with continuation tokens where a `../` could be used if the feature was enabled.

```release-note
API chunking via the `limit` and `continue` request parameters is promoted to beta in this release.  Client libraries using the Informer or ListWatch types will automatically opt in to chunking.
```

smarterclayton committed:
Enable API chunking and promote to beta for 1.9

All list watchers default to using chunking.

smarterclayton commented:
/test pull-kubernetes-kubemark-e2e-gce-big

kargakis commented:
@kubernetes/sig-api-machinery-api-reviews

smarterclayton committed:
Fill partial pages on the server rather than forcing client to

The etcd3 storage now attempts to fill partial pages to prevent clients
having to make more round trips (latency from server to etcd is lower
than client to server). The server makes repeated requests to etcd of
the current page size, then uses the filter function to eliminate any
matches. After this change the apiserver will always return full pages,
but we leave the language in place that clients must tolerate it.

Reduces tail latency of large filtered lists, such as viewing pods
assigned to a node.

smarterclayton committed:
Promote continuation token schema to v1

Change the filtering logic to require a leading path and clean that
afterwards.

smarterclayton commented:
/test pull-kubernetes-kubemark-e2e-gce-big

smarterclayton commented:
/test pull-kubernetes-kubemark-e2e-gce-big

krmayankk commented:
is there public documentation on API chunking ?

smarterclayton commented:
Please see the proposal.

krmayankk commented:
thanks @smarterclayton found it. Adding a link here for others to easily know what this PR does https://github.com/kubernetes/community/blob/58b1c30d95719749068497ba35dfe4c64b21aa72/contributors/design-proposals/api-machinery/api-chunking.md

krmayankk commented:
@smarterclayton is chunking supported in client-go ?

krmayankk commented:
ok i think i found that client-go does support chunking which is super useful, now just need to find which version of client-go supports it

smarterclayton committed:
Use watch cache when rv=0 even when limit is set

smarterclayton commented:
/test pull-kubernetes-kubemark-e2e-gce-big

smarterclayton commented:
Ok, falls back to watch cache now.  I think this is ready for review.

caesarxuchao commented:
/assign @jpbetz

smarterclayton commented:
/test

smarterclayton commented:
/test all
/test pull-kubernetes-kubemark-e2e-gce-big

smarterclayton commented:
/test pull-kubernetes-kubemark-e2e-gce-big

smarterclayton commented:
Tail latencies on a this run for density:

```
DELETE  namespace  endpoints               0.449   0.54    3.606
DELETE  namespace  pods                    2.83    4.558   9.313
DELETE  namespace  replicationcontrollers  2.092   3.294   8.742
GET     cluster    clusterrolebindings     0.467   0.609   1.156
GET     cluster    namespaces              0.422   0.734   2.657
GET     cluster    nodes                   0.107   0.134   0.226
GET     namespace  endpoints               0.443   0.676   3.064
GET     namespace  pods                    0.46    0.755   3.363
GET     namespace  replicationcontrollers  0.452   0.581   2.809
GET     namespace  secrets                 0.102   0.421   1.177
GET     namespace  serviceaccounts         0.37    0.429   0.429
GET     namespace  services                0.435   0.61    2.046
LIST    cluster    cronjobs                0.616   1.326   2.365
LIST    cluster    jobs                    0.615   1.182   3.478
LIST    cluster    namespaces              0.893   1.539   2.214
LIST    cluster    nodes                   22.511  24.396  28.188
LIST    cluster    persistentvolumes       0.737   0.887   1.154
LIST    cluster    pods                    0.162   0.162   0.162
LIST    cluster    replicationcontrollers  0.109   0.216   0.216
LIST    cluster    services                0.741   1.782   1.85
LIST    namespace  limitranges             0.651   1.834   3.524
LIST    namespace  pods                    10.892  11.72   31.21
LIST    namespace  resourcequotas          0.663   1.348   2.153
PATCH   cluster    clusterrolebindings     1.317   1.518   2.345
PATCH   cluster    nodes                   1.576   2.41    5.028
POST    cluster    namespaces              0.836   0.906   0.906
POST    namespace  pods                    1.046   2.248   7.463
POST    namespace  pods                    1.258   2.066   6.892
POST    namespace  replicationcontrollers  1.19    1.873   5.588
POST    namespace  secrets                 2.438   2.692   2.692
POST    namespace  serviceaccounts         0.932   1.017   1.017
PUT     cluster    apiservices             0.776   1.033   2.54
PUT     namespace  endpoints               1.143   2.113   5.338
PUT     namespace  pods                    1.03    2.038   4.603
PUT     namespace  replicationcontrollers  1.028   1.652   3.792
PUT     namespace  serviceaccounts         0.982   1.17    1.17
```

Tail latencies on http://gcsweb.k8s.io/gcs/kubernetes-jenkins/logs/ci-kubernetes-kubemark-500-gce/8733/artifacts/:

```
DELETE  namespace  pods                    4.928   8.793   17.317
DELETE  namespace  replicationcontrollers  3.294   5.161   9.971
GET     cluster    clusterrolebindings     0.802   1.209   3.883
GET     cluster    namespaces              0.756   1.179   4.847
GET     cluster    nodes                   0.186   0.265   0.766
GET     namespace  endpoints               0.734   1.154   3.746
GET     namespace  pods                    0.777   1.417   5.515
GET     namespace  replicationcontrollers  0.698   1.015   3.677
GET     namespace  secrets                 0.178   0.685   2.482
GET     namespace  serviceaccounts         0.555   0.864   0.864
GET     namespace  services                0.6     1.066   2.101
LIST    cluster    cronjobs                0.927   2.418   3.55
LIST    cluster    endpoints               0.164   0.177   0.177
LIST    cluster    jobs                    0.975   2.913   5.839
LIST    cluster    namespaces              1.328   2.78    3.411
LIST    cluster    nodes                   33.394  45.122  91.36
LIST    cluster    persistentvolumes       0.992   2.075   3.374
LIST    cluster    pods                    0.194   0.194   0.194
LIST    cluster    replicationcontrollers  0.181   0.203   0.203
LIST    cluster    services                1.039   2.579   3.13
LIST    namespace  limitranges             1.005   2.012   4.238
LIST    namespace  pods                    13.135  15.309  34.751
LIST    namespace  resourcequotas          1.184   2.839   3.38
PATCH   cluster    clusterrolebindings     1.975   2.669   3.285
PATCH   cluster    nodes                   2.663   4.423   9.368
POST    cluster    namespaces              2.33    2.492   2.492
POST    namespace  pods                    1.779   4.156   17.696
POST    namespace  pods                    2.005   4.311   15.629
POST    namespace  replicationcontrollers  1.936   3.181   5.361
POST    namespace  secrets                 4.293   5.09    5.09
POST    namespace  serviceaccounts         2.097   2.395   2.395
PUT     cluster    apiservices             1.076   1.746   3.196
PUT     namespace  endpoints               1.618   3.209   8.196
PUT     namespace  pods                    1.769   3.792   11.106
PUT     namespace  replicationcontrollers  1.635   2.626   5.929
PUT     namespace  serviceaccounts         1.854   1.961   1.961
```

smarterclayton commented:
Graphed density against 3 recent runs https://docs.google.com/spreadsheets/d/1ooBWLFmJVyzkr4L31SIHEqHTLoHQT0olSC7FtJ30rn0/edit?usp=sharing

99th tail latency is down almost across the board on this particular run.  Will run more to see if this is just a fluke

smarterclayton commented:
/test pull-kubernetes-kubemark-e2e-gce-big

smarterclayton commented:
@kubernetes/sig-scalability-pr-reviews added a second data point - 25% or so latency reduction across the board on kubemark-500

liggitt commented:
Should check with non-kubemark scale test as well

krzyzacy commented:
/retest

smarterclayton commented:
/test pull-kubernetes-kubemark-e2e-gce-big

smarterclayton commented:
@kubernetes/sig-testing-misc https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/52949/pull-kubernetes-e2e-gce-etcd3/57209/?log#log had a test failure but reported success in the CI output

ixdy commented:
@smarterclayton we set `GINKGO_TOLERATE_FLAKES=y` on PR jobs, so all tests get re-run if they fail once and only fail the run if they fail both times.

smarterclayton commented:
Ok, sorry, missed that announcement

smarterclayton commented:
Excellent, I was able to reproduce an issue OpenShift was seeing with watch cache disabled in this run https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/52949/pull-kubernetes-e2e-gce-etcd3/57255/artifacts/e2e-57255-master/kube-apiserver.log

List calls to the API server timeout against etcd 3.1.x for longer than 60s:

```
Trace[254194311]: [1m12.484197726s] [1m12.484192066s] About to write a response                                                                                                  
E1003 00:52:37.908802       5 status.go:62] apiserver received an error that is not an metav1.Status: http: Handler timeout                                                      
E1003 00:52:37.909036       5 writers.go:143] apiserver was unable to write a JSON response: http: Handler timeout                                                               
I1003 00:52:37.911932       5 trace.go:76] Trace[1124568657]: "List /api/v1/namespaces/e2e-tests-replicaset-2csdw/serviceaccounts" (started: 2017-10-03 00:51:26.027164841 +0000 
Trace[1124568657]: [1m11.880462423s] [1m11.88045551s] Listing from storage done                                                                                                  
E1003 00:52:37.913165       5 status.go:62] apiserver received an error that is not an metav1.Status: http: Handler timeout                                                      
E1003 00:52:37.913808       5 writers.go:143] apiserver was unable to write a JSON response: http: Handler timeout                                                               
E1003 00:52:37.914246       5 writers.go:143] apiserver was unable to write a JSON response: http: Handler timeout                                                               
```

smarterclayton commented:
@liggitt i think i'm ok with testing at this point - kubemark big is so finely tuned that it's difficult to get equivalent numbers with watch cache disabled.  I think a follow up item generated by our discussion is:

* should the watch cache allow limiting (it can't do consistent listing) for certain use cases?

smarterclayton commented:
/test pull-kubernetes-kubemark-e2e-gce-big

smarterclayton commented:
ping @kubernetes/sig-api-machinery-pr-reviews PTAL

sttts commented on staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go:
```diff
@@ -429,25 +429,26 @@ func decodeContinue(continueValue, keyPrefix string) (fromKey string, rv int64,
 	}
 	switch c.APIVersion {
-	case "v1alpha1":
+	case "meta.k8s.io/v1":
```
we have constants for this

smarterclayton replied:
These are not publicly accessible - even if we have constants I think we reserve the right to change this.

sttts commented on staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go:
```diff
@@ -494,68 +506,94 @@ func (s *store) List(ctx context.Context, key, resourceVersion string, pred stor
 
+	case s.pagingEnabled && pred.Limit > 0:
+		rangeEnd := clientv3.GetPrefixRangeEnd(keyPrefix)
+		options = append(options, clientv3.WithRange(rangeEnd))
```
missing the RV parsing in this case.

smarterclayton committed:
Avoid intermediate List allocations as items added to the list

Pick a reasonable middle ground between allocating larger chunks of
memory (2048 * ~500b for pod slices) and having many small allocations
as the list is resized by preallocating capacity based on the expected
list size. At worst, we'll allocate a 1M slice for pods and only add
a single pod to it (if the selector is very specific).

liggitt commented:
one comment, lgtm otherwise

smarterclayton commented on staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go:
```diff
@@ -494,68 +507,94 @@ func (s *store) List(ctx context.Context, key, resourceVersion string, pred stor
 		}
 
-		options = append(options, clientv3.WithRange(clientv3.GetPrefixRangeEnd(key)))
+		rangeEnd := clientv3.GetPrefixRangeEnd(keyPrefix)
```
I can add that, better to be restrictive.

smarterclayton commented:
Added more restrictive control on resource version with tests, and also turned the error type into bad request so that clients know they used an invalid input.

smarterclayton commented:
/test pull-kubernetes-kubemark-e2e-gce-big

smarterclayton commented:
/retest

smarterclayton commented:
All green, label on, kubemark continues to show improvement

lavalamp commented on staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go:
```diff
@@ -481,81 +489,166 @@ func (s *store) List(ctx context.Context, key, resourceVersion string, pred stor
+		return
+	}
+	if v.Len() > 0 {
+		extra := reflect.MakeSlice(v.Type(), 0, max)
```
Bug here fixed in #110652. Nearly gets the five year trophy! 🏆
